akka-actor/src/main/scala/akka/actor/Timers.scala

```diff
@@ -0,0 +1,119 @@
+/**
+ * Copyright (C) 2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.actor
+
+import scala.concurrent.duration.FiniteDuration
+import akka.annotation.DoNotInherit
+import akka.util.OptionVal
+
+/**
+ * Scala API: Mix in Timers into your Actor to get support for scheduled
+ * `self` messages via [[TimerScheduler]].
+ *
+ * Timers are bound to the lifecycle of the actor that owns it,
+ * and thus are cancelled automatically when it is restarted or stopped.
+ */
+trait Timers extends Actor {
+
+  private val _timers = new TimerSchedulerImpl(context)
+
+  /**
+   * Start and cancel timers via the enclosed `TimerScheduler`.
+   */
+  final def timers: TimerScheduler = _timers
+
+  override protected[akka] def aroundPreRestart(reason: Throwable, message: Option[Any]): Unit = {
+    timers.cancelAll()
+    super.aroundPreRestart(reason, message)
+  }
+
+  override protected[akka] def aroundPostStop(): Unit = {
+    timers.cancelAll()
+    super.aroundPostStop()
+  }
+
+  override protected[akka] def aroundReceive(receive: Actor.Receive, msg: Any): Unit = {
+    msg match {
+      case timerMsg: TimerSchedulerImpl.TimerMsg ⇒
+        _timers.interceptTimerMsg(timerMsg) match {
+          case OptionVal.Some(m) ⇒ super.aroundReceive(receive, m)
+          case OptionVal.None    ⇒ // discard
+        }
+      case _ ⇒
+        super.aroundReceive(receive, msg)
+    }
+  }
+
+}
+
+/**
+ * Java API: Support for scheduled `self` messages via [[TimerScheduler]].
+ *
+ * Timers are bound to the lifecycle of the actor that owns it,
+ * and thus are cancelled automatically when it is restarted or stopped.
+ */
+abstract class AbstractActorWithTimers extends AbstractActor with Timers {
+  /**
+   * Start and cancel timers via the enclosed `TimerScheduler`.
+   */
+  final def getTimers: TimerScheduler = timers
+}
+
+/**
+ * Support for scheduled `self` messages in an actor.
+ * It is used by mixing in trait `Timers` in Scala or extending `AbstractActorWithTimers`
+ * in Java.
+ *
+ * Timers are bound to the lifecycle of the actor that owns it,
+ * and thus are cancelled automatically when it is restarted or stopped.
+ *
+ * `TimerScheduler` is not thread-safe, i.e. it must only be used within
+ * the actor that owns it.
+ */
+@DoNotInherit abstract class TimerScheduler {
+
+  /**
+   * Start a periodic timer that will send `msg` to the `self` actor at
+   * a fixed `interval`.
+   *
+   * Each timer has a key and if a new timer with same key is started
+   * the previous is cancelled and it's guaranteed that a message from the
+   * previous timer is not received, even though it might already be enqueued
+   * in the mailbox when the new timer is started.
+   */
+  def startPeriodicTimer(key: Any, msg: Any, interval: FiniteDuration): Unit
```

Comment:
that is very rarely used, as far as I know, and with this it's easy to first schedule a single timer and then the period after that, if different initial delay is needed